### PR TITLE
[JAX] Fix get_seqlens_and_offsets() to accept vmapped seg ids and non vmapped seg offsets

### LINKED ISSUE #2685: JAX vmap issue with TE Attention
**Describe the bug**

using JAX vmap results in extra dimension between segment ids and segment pos if initiated with the defaults of the library. This can by bypassed by explicitly passing segment_pos on the user side.

**Steps/Code to reproduce bug**
using fix_vmap=True works
fails when fix_vmap=False:

def _flash_gpu_attention(
        query: jax.Array,
        key: jax.Array,
        value: jax.Array,
        q_segment_ids: Optional[jax.Array],
        kv_segment_ids: Optional[jax.Array],
        scale_factor: float,
        block_sizes: Dict[str, Any],
        fix_vmap: bool,
    ):
        # So this imports won't hurt TPU case
        from transformer_engine.jax.flax.transformer import DotProductAttention  # pytype: disable=import-error
        from transformer_engine.jax.attention import SequenceDescriptor  # pytype: disable=import-error

        if q_segment_ids is None and kv_segment_ids is None:
            segment_ids = None
        else:
            batch_size = query.shape[0]
            q_seq = query.shape[2]
            kv_seq = key.shape[2]
            if q_segment_ids is None and kv_segment_ids is not None:
                q_segment_ids = jnp.ones((batch_size, q_seq), dtype=jnp.int32)
            if kv_segment_ids is None and q_segment_ids is not None:
                kv_segment_ids = jnp.ones((batch_size, kv_seq), dtype=jnp.int32)
            
            segment_pos = None
            if fix_vmap:
                def generate_default_pos(segment_ids):
                    return jnp.where(segment_ids > 0, jnp.arange(segment_ids.shape[-1]), 0)
                
                q_segment_pos = generate_default_pos(q_segment_ids)
                kv_segment_pos = generate_default_pos(kv_segment_ids)
                segment_pos = (q_segment_pos, kv_segment_pos)
                
            segment_ids = SequenceDescriptor.from_segment_ids_and_pos(
                segment_ids=(q_segment_ids, kv_segment_ids),
                segment_pos=segment_pos,
            )

        head_dim = query.shape[-1]

        with nvidia_ctx():
            dpa_layer = DotProductAttention(
                head_dim=head_dim,
                num_attention_heads=query.shape[1],
                num_gqa_groups=key.shape[1],
                attn_mask_type="no_mask",  # 'no_mask', 'padding', 'causal', or 'padding_causal'
                attn_bias_type="no_bias",  # 'no_bias', 'pre_scale_bias' or 'post_scale_bias'
                attention_dropout=0,
                dropout_rng_name="aqt",
                dtype=jax.numpy.bfloat16,
                scale_factor=scale_factor,
                **block_sizes,
            )
            query = query.swapaxes(1, 2).astype(jax.numpy.bfloat16)
            key = key.swapaxes(1, 2).astype(jax.numpy.bfloat16)
            value = value.swapaxes(1, 2).astype(jax.numpy.bfloat16)

            output: jax.Array = dpa_layer.apply({}, query, key, value, segment_ids)
            output = output.swapaxes(1, 2)
            return output

**Expected behavior**

We dont expect a crash when segment pos are not explicitly supplied by the user.

Currently the code crashes at:
.venv/lib/python3.12/site-packages/transformer_engine/jax/attention.py", line 700, in get_seqlens_and_offsets
        assert q_segment_ids.shape == q_segment_pos.shape
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Environment overview (please complete the following information)**

 - Environment location: [Docker]
 - Method of Transformer Engine install: [pip install]. Please specify exact commands you used to install.

Command to install using uv manager:
 uv sync --no-install-project

Project has these dependencies:
dependencies = [
    "jax[cuda12,k8s]==0.9.0.1",
    "jax-cuda12-plugin",
    "jax-cuda12-pjrt",
    "jaxlib",
    "flax==0.12.1",
    "triton; sys_platform == 'linux'", # every group use different version according to other dependencies
    "transformer-engine-cu12>=2.11.0",
    "transformer-engine-jax>=2.11.0",
    "transformer-engine>=2.11.0",

 - If method of install is [Docker], provide `docker pull` & `docker run` commands used

Base docker: nvidia/cuda:12.8.1-devel-ubuntu22.04

Install following dependencies:
sudo apt-get update && sudo apt-get install -y -q --fix-missing \
    libnccl-dev \
    libcudnn9-dev-cuda-12 \
    libcudnn9-cuda-12 

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
ax:    0.9.0.1
jaxlib: 0.9.0.1
numpy:  2.4.2
python: 3.12.12 (main, Oct 14 2025, 21:25:31) [Clang 20.1.4 ]
device info: NVIDIA H100 80GB HBM3-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='h100-1-7b5b844db4-7zdng', release='6.6.97+', version='#1 SMP Fri Aug 22 11:53:37 UTC 2025', machine='x86_64')
XLA_PYTHON_CLIENT_ALLOCATOR=platform
XLA_PYTHON_CLIENT_MEM_FRACTION=0.75
XLA_PYTHON_CLIENT_PREALLOCATE=false

**Device details**
- GPU model - H100 but applicable for other datacenter GPUs like GB200

**Additional context**

### PR DESCRIPTION
**Pull request by KshitijLakhani**
# Description

###  What is the bug ?
TE provides a convenience function `from_segment_ids_and_pos()` which allows users to pass only segment ids and the function returns a `SequenceDescriptor` with internally generated segment pos and passed segment ids.

As mentioned in Issue #2685 , if a user were to vmap a function forward() which i) accepts the q,k,v,segment ids and then ii) calls `from_segment_ids_and_pos()` followed by iii) a call to `DPA()`, what happens is that JAX sees the segment ids as vmapped hence an extra leading dimension is added (e.g. 1,2,128)  whereas the segment offsets are not given a leading dimension (e.g. 2,128). This results in the FusedAttn primitive impl() assert being triggered due to a shape mismatch between seg ids and seg pos as mentioned in issue #2685

### What is the root cause for the bug ?
On debugging, it can be seen that the shape starts differing when the batcher is being traced for the FusedAttn primitive.
`segment_ids` in the primitive: treated as vmapped inputs hence batched → (1, 2, 128).
`segment_pos` in the primitive: treated as derived within the function hence not batched → (2, 128).

Fixes #2685 

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

There are two possible approaches to solve this: 

1.  Ensure that the issue is resolved at the source, i.e.. ensure that  `segment_pos` has the same leading batching dims as `segment_ids`. Add any additional dims in the batcher for the same so that when impl() sees the shape they are the same.  Pros: Issue resolved in a "JAX" way and at source. Cons: Increasing mem be expanding seg pos dims.
2. Resolve the issue when impl() is called, i.e. accomodate for mismatched seg id and seg pos dims when generating the seqlens and offsets. Pros: No extra mem needed as no expansion of dims.  Cons: Not "truely" solved (at source)
 
Second approach is chosen here as it more optimized. After this PR merge the end user can vmap wrap the TE API calls without worrying about the batching in TE.
Accomodate for 

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 1]
```diff
@@ -624,12 +624,58 @@ def convert_to_2d(offsets, batch, max_seqlen):
...
         q_bdim, _, _, _, _, seed_bdim, *_ = batch_dims
 
+        # Ensure segment_pos are batched like segment_ids so impl sees matching shapes.
+        # JAX may give segment_ids batch_dim=0 (i.e. batched) and segment_pos batch_dim=None (i.e. not batched) when
+        # segment_pos were generated inside a vmapped function (e.g. single or nested vmap).
+        batched_args_list = list(batched_args)
+        seg_id_bdim = batch_dims[FusedAttnFwdPrimitive._SEGMENT_IDS_BATCH_DIMS_IDX[0]]
+        seg_pos_bdim = batch_dims[FusedAttnFwdPrimitive._SEGMENT_POS_BATCH_DIMS_IDX[0]]
+        # Expand the segment_pos batch dim to match the segment_ids batch dim, if no batch dim exists for segment_pos
+        if seg_id_bdim is not None and seg_pos_bdim is None and batched_args_list[10].size > 0:
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply transformer_engine/jax/cpp_extensions/attention.py]: This is mostly overkill a check for both q and kv.
Just checking q should suffice (which is also just a sanity guard)

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 2]
```diff
@@ -1079,12 +1125,55 @@ def convert_to_2d(offsets, batch, max_seqlen):
...
         q_bdim, k_bdim, v_bdim, bias_bdim, softmax_offset_bdim, *_ = batch_dims
 
+        # Ensure segment_pos are batched like segment_ids so impl sees matching shapes.
+        # JAX may give segment_ids batch_dim=0 (i.e. batched) and segment_pos batch_dim=None (i.e. not batched) when
+        # segment_pos were generated inside a vmapped function (e.g. single or nested vmap).
+        batched_args_list = list(batched_args)
+        seg_id_bdim = batch_dims[FusedAttnBwdPrimitive._SEGMENT_IDS_BATCH_DIMS_IDX[0]]
+        seg_pos_bdim = batch_dims[FusedAttnBwdPrimitive._SEGMENT_POS_BATCH_DIMS_IDX[0]]
+        # Expand the segment_pos batch dim to match the segment_ids batch dim, if no batch dim exists for segment_pos
+        if seg_id_bdim is not None and seg_pos_bdim is None and batched_args_list[13].size > 0:
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply transformer_engine/jax/cpp_extensions/attention.py]: This is mostly overkill a check for both q and kv.
Just checking q should suffice (which is also just a sanity guard)

**Comment by KshitijLakhani** [comment]
CI passes. The only one failure is due to HF requests for the A100 L2 test.
Rerunning passes these

File: transformer_engine/jax/attention.py [hunk 3]
```diff
@@ -693,26 +695,87 @@ def get_seqlens_and_offsets(
...
+                extra_flat_batch_size_kv = jnp.prod(extra_batch_shape_kv)
+                # vmap below requires same batch size on axis 0 for q_flat and kv_flat; JAX will raise if they differ.
+                q_flat = q_segment_ids.reshape(
+                    extra_flat_batch_size_q, *q_segment_ids.shape[n_extra_batch_dims_q:]
+                )
+                kv_flat = kv_segment_ids.reshape(
+                    extra_flat_batch_size_kv, *kv_segment_ids.shape[n_extra_batch_dims_kv:]
+                )
+
+                def single_extra_batch(seg_id_q, seg_id_kv, seg_pos_q, seg_pos_kv):
```
Comment (jberchtold-nvidia) [inline transformer_engine/jax/attention.py]: Self-Resolve: Is `single_extra_batch` only for capturing the additional static args? If so, what do you think of 

```python
single_extra_batch = functools.partial(_segment_ids_pos_to_seqlens_offsets, attn_mask_type=attn_mask_type, window_size=window_size, max_segments_per_seq=max_segments_per_seq)
```
to make this more concise and not require nested scopes?

Comment (KshitijLakhani) [reply transformer_engine/jax/attention.py]: Makes sense. Thanks for catching
Made the change here: https://github.com/NVIDIA/TransformerEngine/pull/2692/changes/0129045082141c7d15a3d3864f000df845d7adfc
